### DISCUSSION TITLE
カスタムエンジンの構成について

### DISCUSSION THREAD

**Discussion by sevenc-nanashi** [VOICEVOX/voicevox#840] [Ideas]
（投げるのがここでよかったかは判りません。違ったらごめんなさい。）
Voicevoxのカスタムエンジンの仕様について自分なりに考えてみました。参考にしていただければ幸いです。

- `<voicevoxのインストール先>/engines`フォルダを追加する
- `engines`フォルダではエンジン毎にフォルダが別れている（例：`voicevox_root/engines/my_custom`、`voicevox_root/engines/other_custom`）
- エンジンのフォルダは`manifest.json`を持っている必要がある
- `manifest.json`はエンジンの情報を管理する
```ts
interface Manifest {
  // 名前
  name: string

  // ホームページのURL
  url?: string

  // バージョン
  version: string

  // 起動コマンド（Windowsなら`cmd /c`、Macとかだと`/bin/sh -c`で実行する
  command: string

  // HTTPサーバーのポート
  port: number

  // ドキュメントの名前とファイル名
  documents: { [key: string]: string }

  // 開発者の情報
  developer: {
    // 名前
    name: string

    // URL
    url?: string
  } | string  // 名前だけ指定
}
```
例：
```json
{
  "name": "すごそうなカスタムエンジン",
  "url": "https://github.com/sevenc-nanashi/my-engine",
  "version": "1.0.0",
  "command": "./start.exe",
  "port": 12345,
  "documents": {
    "利用規約": "docs/tos.md",
    "ガイドライン": "docs/guideline.md"
  },
  "developer": {
    "name": "名無し｡",
    "url": "https://sevenc7c.com"
  },
}
```

- サーバーは`GET /compatibility`エンドポイントを実装している必要がある
```ts
interface Compatibility {
  cpu: boolean
  cuda: boolean

  accent: boolean
  intonation: boolean
  length: boolean

  cancel_generation: boolean
}
```

## 課題

- 名前問題。speakerId等の衝突が避けられなさそう。
  - `voicevox.default.zundamon` `sevenc7c.my_engine.nanatsuki` のように、`作者ID.エンジンID.キャラID`のような文字列にしてしまう？
     - 移行が大変そう
- 辞書の管理。辞書は現状はvoicevox_engine側に保存している（はず）ので、その後に追加されたエンジンに辞書を渡すことが出来ない？
  - 起動時にデフォエンジンから辞書を取得し、カスタムサーバーの`/dict_init`あたりに投げる？
- ポートの衝突
  - `Manifest.port`ではなく、argvとして渡す？
  - またはcommandに`"./start.exe {port}"`のように`{port}`を入れる？

**Comment by Hiroshiba** [comment]
議論作成ありがとうございます！
カスタムエンジンは僕たちでもかなり重要度の高い、かつ非常に面白い取り組みだと考えています。

実は我々の間では「プラグイン」というユビキタス言語で、似たようなことを進めています！！！

- https://github.com/VOICEVOX/voicevox_project/issues/2

構成も提案いただいた形にかなり似ていて、Manifestなど準備したりしていますが、まだちゃんと構成が固まっていないというステートです。
頂いた提案・課題はとても的を射ていて素晴らしいと感じました！！
もしよかったら↑の議論も組み合わせて、またご提案頂けると非常に嬉しいです。
能力のある方にぜひぜひ進めて頂けるととても心強いです･･･！！

> speakerId等の衝突

すべてUUIDにしてしまう手を考えていて、実際speakersを叩くと話者ごとにUUIDが降ってきます。
が、仰る通り移行がなかなかに大変で、未だ道半ばという感じです。

- https://github.com/VOICEVOX/voicevox/discussions/431#discussioncomment-1578982

> 辞書

こちらもどこかに議論が上がっていて、「VOICEVOX ENGINEを信用して他のエンジンに同期する」みたいなちょっと雑な解決策を考えてたりします。

> ポートの衝突

ポート番号を一元管理できるとエンジン側に持たせても良いと思いますが、VOICEVOX側がそれを管理するのはコストがかかってしまいます。
提案されている通り、ポート番号を設定可能にして、衝突をユーザー側で回避できる仕組みが良いかなと思ってます。
それとは別にデフォルトのポートが分かれてるのは便利そうなので、ある程度はお互い避けるようなのが良いのかなと思います。

---

VOICEVOX discordにて開発者チャンネルがあるので、よかったらどうぞ！
https://twitter.com/hiho_karuta/status/1432554885715677184

**Comment by sevenc-nanashi** [comment]
返信ありがとうございます！
PRなど見てみます。

> ポートの衝突をユーザー側で回避できる仕組み

エンジンごとの設定みたいなのを追加して、そこに色々渡す、が汎用性とかも良さそうかもしれません。
（Manifest.settingsがあっても良さそうかも？[vscodeみたいに](https://code.visualstudio.com/api/references/contribution-points#contributes.configuration)）

#657 にもつながってきそうです。

**Reply by Hiroshiba** [reply]
いいと思います！！
たしか、設定UI内でポートとエンジンのディレクトリを指定可能にする、みたいなことを考えていたと思います。

提案や質問などなんでも聞いていただければ…！！

**Comment by sevenc-nanashi** [comment]
そういえばこんなのありましたね…とりあえず実装されてたのでCLose
